orchagent/aclorch.cpp

```diff
@@ -0,0 +1,1300 @@
+{
+    { IP_TYPE_ANY,         SAI_ACL_IP_TYPE_ANY },
+    { IP_TYPE_IP,          SAI_ACL_IP_TYPE_IP },
+    { IP_TYPE_NON_IP,      SAI_ACL_IP_TYPE_NON_IP },
+    { IP_TYPE_IPv4ANY,     SAI_ACL_IP_TYPE_IPv4ANY },
+    { IP_TYPE_NON_IPv4,    SAI_ACL_IP_TYPE_NON_IPv4 },
+    { IP_TYPE_IPv6ANY,     SAI_ACL_IP_TYPE_IPv6ANY },
+    { IP_TYPE_NON_IPv6,    SAI_ACL_IP_TYPE_NON_IPv6 },
+    { IP_TYPE_ARP,         SAI_ACL_IP_TYPE_ARP },
+    { IP_TYPE_ARP_REQUEST, SAI_ACL_IP_TYPE_ARP_REQUEST },
+    { IP_TYPE_ARP_REPLY,   SAI_ACL_IP_TYPE_ARP_REPLY }
+};
+
+inline string toUpper(const string& str)
+{
+    string uppercase = str;
+
+    transform(uppercase.begin(), uppercase.end(), uppercase.begin(), ::toupper);
+
+    return uppercase;
+}
+
+AclRule::AclRule(AclOrch *aclOrch, string rule, string table) :
+        m_pAclOrch(aclOrch),
+        m_id(rule),
+        m_tableId(table),
+        m_tableOid(SAI_NULL_OBJECT_ID),
+        m_ruleOid(SAI_NULL_OBJECT_ID),
+        m_counterOid(SAI_NULL_OBJECT_ID),
+        m_priority(0)
+{
+    m_tableOid = aclOrch->getTableById(m_tableId);
+}
+
+bool AclRule::validateAddPriority(string attr_name, string attr_value)
+{
+    bool status = false;
+
+    if (attr_name == RULE_PRIORITY)
+    {
+        sai_attribute_t attrs[] =
+        {
+            { SAI_SWITCH_ATTR_ACL_ENTRY_MINIMUM_PRIORITY },
+            { SAI_SWITCH_ATTR_ACL_ENTRY_MAXIMUM_PRIORITY }
+        };
+        // get min/max allowed priority
+        if (sai_switch_api->get_switch_attribute(sizeof(attrs)/sizeof(attrs[0]), attrs) == SAI_STATUS_SUCCESS)
+        {
+            char *endp = NULL;
+            errno = 0;
+            m_priority = strtol(attr_value.c_str(), &endp, 0);
+            // chack conversion was successfull and the value is within the allowed range
+            status = (errno == 0) &&
+                     (endp == attr_value.c_str() + attr_value.size()) &&
+                     (m_priority >= attrs[0].value.u32) &&
+                     (m_priority <= attrs[1].value.u32);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Failed to get ACL entry priority min/max values");
+        }
+    }
+
+    return status;
+}
+
+bool AclRule::validateAddMatch(string attr_name, string attr_value)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_value_t value;
+
+    if (aclMatchLookup.find(attr_name) == aclMatchLookup.end())
+    {
+        return false;
+    }
+    else if(attr_name == MATCH_IP_TYPE)
+    {
+        if (!processIpType(attr_value, value.aclfield.data.u32))
+        {
+            SWSS_LOG_ERROR("Invalid IP type %s", attr_value.c_str());
+            return false;
+        }
+
+        value.aclfield.enable = true;
+        value.aclfield.mask.u32 = 0xFFFFFFFF;
+    }
+    else if((attr_name == MATCH_ETHER_TYPE)  || (attr_name == MATCH_DSCP)        ||
+            (attr_name == MATCH_L4_SRC_PORT) || (attr_name == MATCH_L4_DST_PORT) ||
+            (attr_name == MATCH_IP_PROTOCOL) || (attr_name == MATCH_TCP_FLAGS))
```

Comment:
MATCH_TCP_FLAGS needs to have specified mask in the swssconfig JSON file.
Because we will need to match single bit flags, e.g. only match ACK flags including SYN/ACK, FIN/ACK, ACK, etc.